jgrapht-core/src/main/java/org/jgrapht/alg/isomorphism/AHUForestIsomorphismInspector.java

```diff
@@ -0,0 +1,236 @@
+     * @throws IllegalArgumentException if {@code roots2} is empty
+     * @throws IllegalArgumentException if either {@code roots1} or {@code roots2} contain an invalid vertex
+     */
+    public AHUForestIsomorphismInspector(Graph<V, E> forest1, Set<V> roots1, Graph<V, E> forest2, Set<V> roots2){
+        this.forest1 = Objects.requireNonNull(forest1, "input forest cannot be null");
+        this.forest2 = Objects.requireNonNull(forest2, "input forest cannot be null");
+
+        this.roots1 = Objects.requireNonNull(roots1, "set of roots cannot be null");
+        this.roots2 = Objects.requireNonNull(roots2, "set of roots cannot be null");
+
+        if (roots1.isEmpty()){
+            throw new IllegalArgumentException("roots1 cannot be empty");
+        }
+
+        if (roots2.isEmpty()){
+            throw new IllegalArgumentException("roots2 cannot be empty");
+        }
+
+        if (!forest1.vertexSet().containsAll(roots1)){
+            throw new IllegalArgumentException("root not contained in forest");
+        }
+
+        if (!forest2.vertexSet().containsAll(roots2)){
+            throw new IllegalArgumentException("root not contained in forest");
+        }
+    }
+
+    private V getFreshVertex(Graph<V, E> graph){
+        Supplier<V> supplier = graph.getVertexSupplier();
+
+        if (Objects.isNull(supplier))
+            throw new IllegalArgumentException("vertex supplier cannot be null");
+
+        V v = supplier.get();
+        assert !graph.vertexSet().contains(v);
+        return v;
+    }
+
+    private Graph<V, E> takeSubgraph(Graph<V, E> graph, Set<V> vertices){
+        Graph<V, E> subgraph = new SimpleGraph<>(graph.getVertexSupplier(), graph.getEdgeSupplier(), false);
+
+        for (V v: vertices)
+            subgraph.addVertex(v);
+
+        for (V v: vertices){
+            for (E edge: graph.edgesOf(v)){
+                V u = Graphs.getOppositeVertex(graph, edge, v);
+
+                if (vertices.contains(u))
+                    subgraph.addEdge(u, v);
+            }
+        }
+
+        return subgraph;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Iterator<GraphMapping<V, E>> getMappings() {
+        GraphMapping<V, E> iterMapping = getMapping();
+
+        if (iterMapping == null)
+            return Collections.emptyIterator();
+        else
+            return Collections.singletonList(iterMapping).iterator();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isomorphismExists(){
+        if (!computed){
+            isomorphicMapping = getMapping();
+        }
+
+        return isomorphicMapping != null;
+    }
+
+    private Pair<V, Graph<V, E>> addDummyRoot(Graph<V, E> forest, Set<V> roots){
+        Graph<V, E> freshForest =
+                new SimpleGraph<>(forest.getVertexSupplier(), forest.getEdgeSupplier(), false);
+
+        roots.forEach(freshForest::addVertex);
+        V freshVertex = getFreshVertex(forest);
```

Comment:
I think this is now equivalent to just calling addVertex() with no args on freshForest.

Bonus comment:  this method makes me think of those little tree-shaped car air fresheners.

Reply:
I am not sure they are equivalent. If no vertex supplier exists then `getFreshVertex` will detect that and throw an IllegalArgumentException while addVertex() will throw a NullPointerExeption.

Reply:
In #650 we changed addVertex() to throw IllegalArgumentException.

Reply:
You mean UnsupportedOperationException?

https://github.com/jgrapht/jgrapht/blob/53d3fef56ac4fe96839b05a1cf3d275937ba8e3b/jgrapht-core/src/main/java/org/jgrapht/graph/AbstractBaseGraph.java#L274

Either way, I think addVertex() is all you need here.